### PR TITLE
fix: Optimize read_side_padding"

### PR DESCRIPTION
**Pull request by kazuyukitanimura**
## Rationale for this change

This PR improves read_side_padding" that is used for CHAR() schema

## What changes are included in this PR?

Optimized spark_read_side_padding"

## How are these changes tested?

Added tests

### CODE REVIEW

**Comment by kazuyukitanimura** [comment]
### Before
Screenshot 2024-08-03 at 3 43 38 AM

### After
Screenshot 2024-08-03 at 3 43 54 AM

File: spark/src/test/resources/tpcds-micro-benchmarks/char_type.sql [hunk 1]
```diff
@@ -0,0 +1,7 @@
+SELECT
```
Comment (comphead) [inline spark/src/test/resources/tpcds-micro-benchmarks/char_type.sql:1]: how this test related to rpad? 🤔

**Review by comphead — APPROVED** [review]
lgtm thanks @kazuyukitanimura and the benchmark results are promising

File: native/spark-expr/src/scalar_funcs.rs [hunk 2]
```diff
@@ -410,29 +410,31 @@ fn spark_rpad_internal<T: OffsetSizeTrait>(
...
+        match string {
             Some(string) => {
-                let length = if length < 0 { 0 } else { length as usize };
-                if length == 0 {
-                    Ok(Some("".to_string()))
+                // It looks Spark's UTF8String is closer to chars rather than graphemes
+                // https://stackoverflow.com/a/46290728
+                let char_len = string.chars().count();
+                if length <= char_len {
+                    builder.append_value(string);
```
Comment (viirya) [inline native/spark-expr/src/scalar_funcs.rs:426]: If the required len is less than string's length, don't we need to take substring of it? Spark RPad does it.

Comment (viirya) [reply native/spark-expr/src/scalar_funcs.rs:426]: Current implementation already has this issue.

File: native/spark-expr/src/scalar_funcs.rs [hunk 3]
```diff
@@ -410,29 +410,31 @@ fn spark_rpad_internal<T: OffsetSizeTrait>(
...
-        .iter()
-        .map(|string| match string {
+    for string in string_array.iter() {
+        match string {
             Some(string) => {
-                let length = if length < 0 { 0 } else { length as usize };
-                if length == 0 {
-                    Ok(Some("".to_string()))
+                // It looks Spark's UTF8String is closer to chars rather than graphemes
+                // https://stackoverflow.com/a/46290728
```
Comment (viirya) [inline native/spark-expr/src/scalar_funcs.rs:423]: Can you add an unit test for that?

File: spark/src/test/resources/tpcds-micro-benchmarks/char_type.sql (hunk 1 above)
Comment (kazuyukitanimura) [reply spark/src/test/resources/tpcds-micro-benchmarks/char_type.sql:1]: They are related as their schema types are CHAR()

File: native/spark-expr/src/scalar_funcs.rs (hunk 2 above)
Comment (kazuyukitanimura) [reply native/spark-expr/src/scalar_funcs.rs:426]: In the line 389 there is an existing comment
```
/// Similar to DataFusion `rpad`, but not to truncate when the string is already longer than length
```
Perhaps I should change the name of this method, this is not used for rpad

File: native/spark-expr/src/scalar_funcs.rs (hunk 3 above)
Comment (kazuyukitanimura) [reply native/spark-expr/src/scalar_funcs.rs:423]: added

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala [hunk 4]
```diff
@@ -1911,6 +1911,18 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
     }
   }
 
+  test("readSidePadding") {
+    // https://stackoverflow.com/a/46290728
+    val table = "test"
+    withTable(table) {
+      sql(s"create table $table(col1 CHAR(2)) using parquet")
+      sql(s"insert into $table values('é')") // unicode 'e\\u{301}'
+      sql(s"insert into $table values('é')") // unicode '\\u{e9}'
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:1920]: for completeness, could you also test for empty string and 2 char string?

**Review by andygrove — APPROVED** [review]
LGTM. I left a comment about expanding the test

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala (hunk 4 above)
Comment (kazuyukitanimura) [reply spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:1920]: Added

**Comment by kazuyukitanimura** [comment]
Merged, Thanks @comphead @viirya @andygrove
